lib/dealership.rb

```diff
@@ -0,0 +1,86 @@
+class Dealership
+
+  attr_reader :name, :address, :inventory
+
+  def initialize(name, address)
+    @name = name
+    @address = address
+    @inventory = []
+  end
+
+  def inventory_count
+    @inventory.length
+  end
+
+  def has_inventory?
+    @inventory.length >= 1
+  end
+
+  def add_car(car)
+    @inventory << car
+  end
+
+  def cars_by_make(make)
+    cars_by_make_catcher = []
+    @inventory.each do |car|
+      if car.make == make
+        cars_by_make_catcher << car
+      end
+    end
+    cars_by_make_catcher
+  end
+
+  def total_value
+    total_inventory_value = 0
+    @inventory.each do |car|
+      total_inventory_value += car.total_cost
+    end
+    total_inventory_value
+  end
+
+  def details
+    dealer_details = {
+      "total_value" => total_value,
+      "address" => @address
+    }
+  end
+
+  def average_price_of_car
+    average_cost = total_value / inventory_count
+    format_number(average_cost)
+  end
+
+  def format_number(number)
+    group_nums = number.to_s.chars.to_a.reverse.each_slice(3) #
+    group_nums.map(&:join).join(',').reverse
+  end
+
+  def cars_sorted_by_price
+    sorted = @inventory.sort_by {|car| car.total_cost}
+  end
+
+  def inventory_hash
+    # Refactored just in time!!
+    inventory_hash = Hash.new
+    @inventory.each do |car|
+      inventory_hash[car.make] = cars_by_make(car.make).map{|car| car}
+    end
+    inventory_hash
```

Comment:
I don't think you have learned about Big O(n) or optimization yet, but just to push you a bit in your learning -- this is a lot of iterations over our inventory. Basically for each car, you are looping through all the cars again. Another way to do this that is more efficient but a little more verbose could be:

```
def inventory_hash
  inventory_hash = Hash.new
  inventory.each do |car|
    if inventory_hash[car.make]
      inventory_hash[car.make] = inventory_hash[car.make] << car
    else
      inventory_hash[car.make] = [car]
    end
  end
end
```

or you can use the group_by enumerator

```
def inventory_hash
  @inventory.group_by { |car| car.make }
end
```